### PR TITLE
user-settable materialization & observation tags

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

Currently, asset materializations and asset observation events have a `tags` field. However, users are not able/expected to set values on this field. This PR changes that.

A notable aspect of what's being proposed here: users can override the system tags, like those we set to track data provenance at the materialization level.

The direct catalyst for this PR was a conversation I had with an ML-focused user, who wanted to be able to customize data provenance tracking. More info here: https://github.com/dagster-io/dagster/issues/21436.

I'm not yet convinced this is the right way to handle that need. An alternative would be something like a `provenance` argument on `MaterializeResult` and `Output`. But this approach involves less surface area in the long term.

### CODE REVIEW

**Review by schrockn — CHANGES_REQUESTED** [review]
Back to your q. Definitely supportive. Let me know when ready for full review

File: python_modules/dagster/dagster/_core/definitions/events.py [hunk 1]
```diff
@@ -88,6 +93,8 @@ class Output(Generic[T]):
             list, and one of the data classes returned by a MetadataValue static method.
         data_version (Optional[DataVersion]): (Experimental) A data version to manually set
             for the asset.
+        tags (Optional[Mapping[str, str]]): (Experimental) Tags that will be attached to the asset
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/events.py:96]: Should this be named something like `asset_event_tags`?

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/events.py:96]: At a minimum, I think it is important for the docstring to say "Tags that will be attached to the asset materialization"

My general feeling is that `tags` is sufficient.

File: python_modules/dagster/dagster/_core/definitions/events.py [hunk 2]
```diff
@@ -690,3 +699,35 @@ def __new__(cls, hook_name: str, is_skipped: Optional[bool] = None):
...
+        # - The value for the data version tag can be an arbitrary string
+        if key in [CODE_VERSION_TAG, DATA_VERSION_TAG, _OLD_DATA_VERSION_TAG]:
+            continue
+
+        if (
+            key.startswith(INPUT_DATA_VERSION_TAG_PREFIX)
+            or key.startswith(INPUT_EVENT_POINTER_TAG_PREFIX)
+            or key.startswith(_OLD_INPUT_DATA_VERSION_TAG_PREFIX)
+        ):
+            continue
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/events.py:729]: can we encapsulate this in a helper function and document it nicely? e.g. (`is_system_defined_asset_event_tag`) or something?

Also what is the stated behavior if a user provides their own version of these tags?

**Review by schrockn — CHANGES_REQUESTED** [review]
Pretty much good to go. My one substantive question is about our behavior when the user specifies their own system tags.

File: python_modules/dagster/dagster/_core/definitions/events.py (hunk 2 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/events.py:729]: > Also what is the stated behavior if a user provides their own version of these tags?

It overrides the system-provided tag.  That's actually the motivation for this PR, to allow overriding the provenance tags.

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/events.py:729]: Got it.

I think some function like

```python
def is_system_asset_event_tag(key):
  return key in [CODE_VERSION_TAG, DATA_VERSION_TAG, _OLD_DATA_VERSION_TAG]
    or key.startswith(INPUT_DATA_VERSION_TAG_PREFIX)
    or key.startswith(INPUT_EVENT_POINTER_TAG_PREFIX)
    or key.startswith(_OLD_INPUT_DATA_VERSION_TAG_PREFIX)
```

would make this much more clear

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/events.py:729]: 👍

**Pull request merged**